### PR TITLE
Use dotnet ForEachAsync implementation

### PR DESCRIPTION
**Pull request by sebastienros**
## Summary by CodeRabbit

- **New Features**
	- Enhanced extension loading by implementing parallel processing, improving performance.
	- Optimized background services by utilizing parallel task execution for shell operations.

### CODE REVIEW

**Comment by sebastienros** [comment]
@MikeAlhayek can you try this branch? It's fixing the issue for me.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:301]:
```suggestion
                await Parallel.ForEach(modules, (module, cancellationToken) =>
```

**Comment by MikeAlhayek** [comment]
@sebastienros this is a good change. But it did not fix the issue in reference.

**Comment by sebastienros** [comment]
@MikeAlhayek 😢

**Comment by sebastienros** [comment]
Fucking rabbit.

**Comment by MikeAlhayek** [comment]
I did not think it would fix the issue here because the only thing that could impact the setup process is the change in the ExtensionManager. but event that, it should not impact it since none of the calls in that parallel logic use database call.

I tried to explain what I think is happening in the comment here https://github.com/OrchardCMS/OrchardCore/issues/15794#issuecomment-2088698684

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:301]: Why not use `Parallel.ForEachAsync()` here too?

**Review by Piedone — CHANGES_REQUESTED** [review]
This does fix both issues, awesome!

**Comment by MikeAlhayek** [comment]
> This does fix both issues, awesome!

It does not fix the issue for me. Strange. Let me try cleaning up my environment and test it out again.

**Comment by Piedone** [comment]
FYI I tried a repro of both issues with a local SQL Server.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:301]: the delegate is synchronous. No need for asynchronous

Comment (Piedone) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:301]: `EnsureInitializedAsync()` is though so it can make use of `async`.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:1]: Also do this, please: https://github.com/OrchardCMS/OrchardCore/issues/15794#issuecomment-2069323599 It doesn't fix the issue for MySQL: https://github.com/OrchardCMS/OrchardCore/actions/runs/8914513161/job/24482248824 and there's some other problem with SQL Server.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:301]: but that is done before the parallel call. The parallel call does not need async

Comment (Piedone) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:301]: `Parallel.ForEach()` blocks the calling thread, since it's synchronous. So, that thread will just wait, like synchronous IO does. `Parallel.ForEachAsync()` being async, the calling thread won't be blocked (and can thus be used to e.g. serve as one of the threads used for the loop).

Comment (sebastienros) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:301]: >  the calling thread won't be blocked (and can thus be used to e.g. serve as one of the threads used for the loop)

But it won't because there is no async code in the parallel lambdas.

**Comment by sebastienros** [comment]
It was failing every time for me before this change, then worked twice in a row, maybe I was lucky? Will continue the investigations on my side. We can still merge it since it's less custom code.

**Comment by MikeAlhayek** [comment]
> It was failing every time for me before this change, then worked twice in a row, maybe I was lucky? Will continue the investigations on my side. We can still merge it since it's less custom code.

Did you try to write to a remote SQL server instead of a local instance? Just so you have higher latency

**Comment by MikeAlhayek** [comment]
Oh and don't step through the code. remove all break points and let the code execute naturally otherwise you'll avoid the concurrency

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs
Comment (Piedone) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:301]: What will the thread of `EnsureInitializedAsync()` do until `Parallel.ForEach()` returns?

**Comment by Piedone** [comment]
With Azure SQL it fails for me as well, sadly, though differently (but MARS also hints on parallel queries):

```
System.InvalidOperationException: The connection does not support MultipleActiveResultSets.
   at Microsoft.Data.SqlClient.SqlInternalConnectionTds.ValidateConnectionForExecute(SqlCommand command)
   at Microsoft.Data.SqlClient.SqlInternalTransaction.Rollback()
   at Microsoft.Data.SqlClient.SqlInternalTransaction.Dispose(Boolean disposing)
   at Microsoft.Data.SqlClient.SqlInternalTransaction.Dispose()
   at Microsoft.Data.SqlClient.SqlTransaction.Dispose(Boolean disposing)
   at System.Data.Common.DbTransaction.DisposeAsync()
   at YesSql.Session.ReleaseTransactionAsync()
   at YesSql.Session.ReleaseConnectionAsync()
   at YesSql.Session.CommitOrRollbackTransactionAsync()
   at YesSql.Session.SaveChangesAsync()
   at OrchardCore.Data.Documents.DocumentStore.CommitAsync() in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Data.YesSql\Documents\DocumentStore.cs:line 121
   at OrchardCore.Environment.Shell.Scope.ShellScope.BeforeDisposeAsync() in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Abstractions\Shell\Scope\ShellScope.cs:line 397
   at OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func`2 execute, Boolean activateShell) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Abstractions\Shell\Scope\ShellScope.cs:line 267
   at OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func`2 execute, Boolean activateShell) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Abstractions\Shell\Scope\ShellScope.cs:line 269
   at OrchardCore.Recipes.Services.RecipeExecutor.ExecuteStepAsync(RecipeExecutionContext recipeStep) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Recipes.Core\Services\RecipeExecutor.cs:line 156
   at OrchardCore.Recipes.Services.RecipeExecutor.ExecuteAsync(String executionId, RecipeDescriptor recipeDescriptor, IDictionary`2 environment, CancellationToken cancellationToken) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Recipes.Core\Services\RecipeExecutor.cs:line 99
   at OrchardCore.Recipes.Services.RecipeExecutor.ExecuteAsync(String executionId, RecipeDescriptor recipeDescriptor, IDictionary`2 environment, CancellationToken cancellationToken) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Recipes.Core\Services\RecipeExecutor.cs:line 117
   at OrchardCore.Recipes.Services.RecipeExecutor.ExecuteAsync(String executionId, RecipeDescriptor recipeDescriptor, IDictionary`2 environment, CancellationToken cancellationToken) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Recipes.Core\Services\RecipeExecutor.cs:line 132
   at OrchardCore.Recipes.Services.RecipeExecutor.ExecuteAsync(String executionId, RecipeDescriptor recipeDescriptor, IDictionary`2 environment, CancellationToken cancellationToken) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Recipes.Core\Services\RecipeExecutor.cs:line 142
   at OrchardCore.Setup.Services.SetupService.SetupInternalAsync(SetupContext context) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Setup.Core\SetupService.cs:line 231
   at OrchardCore.Setup.Services.SetupService.SetupInternalAsync(SetupContext context) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Setup.Core\SetupService.cs:line 232
   at OrchardCore.Setup.Services.SetupService.SetupAsync(SetupContext context) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Setup.Core\SetupService.cs:line 95
   at OrchardCore.Setup.Services.SetupService.SetupAsync(SetupContext context) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Setup.Core\SetupService.cs:line 110
   at OrchardCore.Setup.Controllers.SetupController.IndexPOST(SetupViewModel model) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore.Modules\OrchardCore.Setup\Controllers\SetupController.cs:line 176
   at lambda_method76(Closure, Object)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfActionResultExecutor.Execute(ActionContext actionContext, IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at OrchardCore.Modules.ModularTenantRouterMiddleware.Invoke(HttpContext httpContext) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 47
   at OrchardCore.Modules.ModularTenantContainerMiddleware.<>c__DisplayClass4_0.<<Invoke>b__0>d.MoveNext() in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore\Modules\ModularTenantContainerMiddleware.cs:line 61
--- End of stack trace from previous location ---
   at OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func`2 execute, Boolean activateShell) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Abstractions\Shell\Scope\ShellScope.cs:line 253
   at OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func`2 execute, Boolean activateShell) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Abstractions\Shell\Scope\ShellScope.cs:line 258
   at OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func`2 execute, Boolean activateShell) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Abstractions\Shell\Scope\ShellScope.cs:line 263
   at OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func`2 execute, Boolean activateShell) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Abstractions\Shell\Scope\ShellScope.cs:line 268
   at OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func`2 execute, Boolean activateShell) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore.Abstractions\Shell\Scope\ShellScope.cs:line 269
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in E:\Projects\OrchardForks\OrchardCore5\src\OrchardCore\OrchardCore\Modules\ModularTenantContainerMiddleware.cs:line 59
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)
```

BTW neither previously nor now did I have breakpoints or anything that would inhibit the normal code flow, just running with Ctrl+F5.

**Comment by MikeAlhayek** [comment]
> BTW neither previously nor now did I have breakpoints or anything that would inhibit the normal code flow, just running with Ctrl+F5.

Yes I think the latency is the key here. If you connect locally "very low latency" and it would work. But when you connect to a remote Azure SQL Server which has higher latency, you encounter the issue.

Try to remove `RolesStep` from the IoC and try it.

**Comment by MikeAlhayek** [comment]
> Will continue the investigations on my side. We can still merge it since it's less custom code.

we should merge this as it is a good change. But please remove the referenced 2 issues since it does not really fix them.

**Review by MikeAlhayek — APPROVED** [review]

**Comment by MikeAlhayek** [comment]
@sebastienros here are the last few trace logs from my logfile just before we countered the issue "maybe the queries could give you more clues on what could be updating"

```
024-05-01 15:21:34.2271|ocblog29|00-f43e7e43218ecd49cc90714d6dea92f5-3b1e9d38f7a50749-00||OrchardCore.Recipes.Services.RecipeExecutor|DEBUG|Executing the step themes from blog.recipe.json was successful. 
2024-05-01 15:21:34.2271|ocblog29|00-f43e7e43218ecd49cc90714d6dea92f5-3b1e9d38f7a50749-00||OrchardCore.Recipes.Services.RecipeExecutor|DEBUG|Executing the step Roles from blog.recipe.json 
2024-05-01 15:21:34.2271|ocblog29|00-f43e7e43218ecd49cc90714d6dea92f5-3b1e9d38f7a50749-00||OrchardCore.Recipes.Services.RecipeExecutor|DEBUG|Executing Recipe using a shellScope. RequireNewScope; True 
2024-05-01 15:21:34.2692|ocblog29|00-f43e7e43218ecd49cc90714d6dea92f5-3b1e9d38f7a50749-00||OrchardCore.Recipes.Services.RecipeExecutor|INFO|Executing recipe step 'Roles'. 
2024-05-01 15:21:34.2722|ocblog29|00-f43e7e43218ecd49cc90714d6dea92f5-3b1e9d38f7a50749-00||YesSql|DEBUG|SELECT TOP (1) [ocblog29_Document].* FROM [ocblog29_Document] WHERE [ocblog29_Document].[Type] = @Type 
2024-05-01 15:21:34.3597|ocblog29|00-f43e7e43218ecd49cc90714d6dea92f5-3b1e9d38f7a50749-00||YesSql|DEBUG|SELECT TOP (1) [ocblog29_Document].* FROM [ocblog29_Document] WHERE [ocblog29_Document].[Type] = @Type 
2024-05-01 15:21:34.4523|ocblog29|00-f43e7e43218ecd49cc90714d6dea92f5-3b1e9d38f7a50749-00||YesSql|DEBUG|SELECT TOP (1) [ocblog29_Document].* FROM [ocblog29_Document] WHERE [ocblog29_Document].[Type] = @Type 
2024-05-01 15:21:34.6370|ocblog29|00-f43e7e43218ecd49cc90714d6dea92f5-3b1e9d38f7a50749-00||YesSql|TRACE|insert into [ocblog29_Document] ([Id], [Type], [Content], [Version]) values (@Id_1, @Type_1, @Content_1, @Version_1); 
2024-05-01 15:21:34.6370|ocblog29|00-f43e7e43218ecd49cc90714d6dea92f5-3b1e9d38f7a50749-00||OrchardCore.Recipes.Services.RecipeExecutor|INFO|Finished executing recipe step 'Roles'. 
2024-05-01 15:21:34.7262|ocblog29|00-f43e7e43218ecd49cc90714d6dea92f5-3b1e9d38f7a50749-00||YesSql|DEBUG|SELECT TOP (1) [ocblog29_Document].* FROM [ocblog29_Document] WHERE [ocblog29_Document].[Type] = @Type 
2024-05-01 15:21:35.5918|ocblog29|00-f43e7e43218ecd49cc90714d6dea92f5-3b1e9d38f7a50749-00||YesSql|TRACE|update [ocblog29_Document] set [Content] = @Content_1, [Version] = @Version_1 where [Id] = @Id_1; 
2024-05-01 15:21:35.7158|ocblog29|00-f43e7e43218ecd49cc90714d6dea92f5-3b1e9d38f7a50749-00||YesSql|DEBUG|SELECT TOP (1) [ocblog29_Document].* FROM [ocblog29_Document] WHERE [ocblog29_Document].[Type] = @Type 
2024-05-01 15:21:41.0712|ocblog29|00-f43e7e43218ecd49cc90714d6dea92f5-3b1e9d38f7a50749-00||OrchardCore.Recipes.Services.RecipeExecutor|DEBUG|Executing the step Roles from blog.recipe.json was unsuccessful. Error: Microsoft.Data.SqlClient.SqlException (0x80131904): The transaction operation cannot be performed because there are pending requests working on this transaction.
```

**Comment by Piedone** [comment]
Without `RolesStep` it works indeed. I'm OK with this, once we have https://github.com/OrchardCMS/OrchardCore/pull/15936#discussion_r1586775715 settled.

**Comment by Piedone** [comment]
@sebastienros why did you merge?

**Comment by sebastienros** [comment]
Got an approval, and your comment is not correct.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs
Comment (sebastienros) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:301]: The only other thing we could change is that since nothing is asynchronous we can remove the async pattern on this method. And replace the semaphore with a lock. To remove any assumption that this is actually async.

**Comment by Piedone** [comment]
You asked for my review too, and we were in a conversation, what I was quite clear about I'd like to finish before the merge. So I feel sidestepped that you merged it despite that. It's no problem if we disagree, but if I put an effort into reviewing your PR, as you asked, then I'd expect you to engage with me as well, even if I ultimately prove to be wrong. I'm happy to learn. The goal is to have good code, what also comes from discussions like this. If not by changes to the code in this PR, then the reviewer learning something and doing it better in the next one.

So, can you then please tell what will the thread of `EnsureInitializedAsync()` do until `Parallel.ForEach()` returns if not be blocked? Because it looks to me that now we have synchronous code in an async method that blocks, while we could also have it fully async.

**Comment by sebastienros** [comment]
Sorry, I had not seen this part:

> I'm OK with this, once we have https://github.com/OrchardCMS/OrchardCore/pull/15936#discussion_r1586775715 settled.

With both solutions the caller thread will block until the parallel operation is done because the body in the parallel operation is not async, it will never yield.

**Comment by Piedone** [comment]
Why does it block with `Parallel.ForeachAsync()`? I know that it does a lot-lot more and differently, but it seems to me that it's roughly analogous to:

```c#
var tasks = new List<Task>();

foreach (var module in modules)
{
    tasks.Add(Task.Run(...));
}

await Task.WhenAll(tasks);
```

So, while the body lambda is executed with a given `module` on multiple threads in parallel, the caller is not blocked but waiting asynchronously.

**Comment by sebastienros** [comment]
I don't think it's using Task.WhenAll at all. ThreadPool + TaskCompletionSource to detect all threads are done. Where is the documentation about ForEachAsync that would tell if it should always be used even when there is no async body?

**Comment by Piedone** [comment]
It doesn't use `Task.WhenAll`, what I mean by it being roughly _analogous_ is that it's in behavior similar to the above code. I.e. it runs the body in parallel and completes when all parallel bodies are completed.

I'm not saying `Parallel.ForeachAsync()` should always be used when there's no async body. Rather, that it's more suitable, regardless of the body, in this case, because the consumer is async. The body returning `ValueTask` and not `Task` BTW hints on the authors expecting a large number of body implementations to be sync.

**Pull request merged**
